Firestore/core/test/unit/credentials/firebase_app_check_credentials_provider_test.mm

```diff
@@ -220,6 +220,45 @@ - (nonnull NSString*)tokenDidChangeNotificationName {
   EXPECT_EQ("token2", token_future.get());
 }
 
+// Regression test for https://github.com/firebase/firebase-ios-sdk/issues/8895
+TEST(FirebaseAppCheckCredentialsProviderTest,
+     ListenDoesNotCrashIfUnrelatedNotificationsAreInvalid) {
+  auto token_promise = std::make_shared<std::promise<std::string>>();
+
+  FIRApp* app = testutil::AppForUnitTesting();
+  FSTAppCheckFake* app_check = [[FSTAppCheckFake alloc] initWithToken:@""];
+  FirebaseAppCheckCredentialsProvider credentials_provider(app, app_check);
+
+  credentials_provider.SetCredentialChangeListener(
+      [token_promise](const std::string& result) {
+        if (result != "") {
+          token_promise->set_value(result);
+        }
+      });
+
+  // Sending this notifcation would cause a crash if it was processed in the
+  // AppCheck notification handlder since AppCheck expects the userInfo object
+  // to an NSDictionary.
+  id userInfo = @"this_should_be_a_dictionary";
```

Comment:
Hm, [userInfo](https://developer.apple.com/documentation/foundation/nsnotification/1409222-userinfo?language=objc) is expected to be a dictionary. If it is not then there is a bug in somewhere else on the notification poster side.

Reply:
I guess if we expect `userInfo` to be not a dictionary then we should handle it gracefully in the code. From the other side I'm not sure if we should code defensively against such types of issues. I don't feel strongly about it but this specific test seems to be a bit redundant to me.

Reply:
It's the only way I could reproduce the issue, so while it is a bit redundant it is IMHO worth keeping as it tests the entire problematic code flow, even if the fix is of much narrower scope.